File: sj/49week/Length of Longest Fibonacci Subsequence.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var lenLongestFibSubseq = function(arr) {
    
    let map = new Map();
    
    let len = 0;
    
    for(let a of arr)
        {
            map.set(a,true);
        }
    
    arr.sort((a,b)=> a-b);
    
    
    for(let i=0; i<arr.length; i++)
        {
            for(let j=i+1; j<arr.length; j++)
                {
                    let right = arr[j];
                    
                    let left = arr[i];
                    
                    let templen = 0;
                    
                    //두개 더한게 있는지
                    while(map.has(left+right))
                        {
                            templen++;
                            // 앞에 두개 더한값
                            let temp = left+right;
                            // 앞으로 댕김
                            left= right;
                            
                            right = temp;
                        }
                    
                    len = Math.max(len, templen+2);
                }
        }
    return len>2? len:0;
    
};

File: jw/week50/LengthOfLongestFibonacciSubsequence.java
class Solution {
    public int lenLongestFibSubseq(int[] arr) {
        Set<Integer> st = new HashSet<>();
        for(int i : arr) st.add(i);
        
        int max = 0;
        
        for(int i=0;i<arr.length-1;i++) {
            for(int j=i+1;j<arr.length;j++) {
                int s = 2;
                int ex = arr[j];
                int t = arr[i] + arr[j];
               
                while(st.contains(t)) {
                    s++;
                    int tmp = ex;
                    ex = t;
                    t = tmp + t;
                    max=Math.max(max,s);
                }
            }
        }
        
        return max;
    }
}
